geoh5py/shared/cut_by_extent.py

```diff
@@ -0,0 +1,551 @@
+# ''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+#  Copyright (c) 2025 Mira Geoscience Ltd.                                     '
+#                                                                              '
+#  This file is part of geoh5py.                                               '
+#                                                                              '
+#  geoh5py is free software: you can redistribute it and/or modify             '
+#  it under the terms of the GNU Lesser General Public License as published by '
+#  the Free Software Foundation, either version 3 of the License, or           '
+#  (at your option) any later version.                                         '
+#                                                                              '
+#  geoh5py is distributed in the hope that it will be useful,                  '
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of              '
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the               '
+#  GNU Lesser General Public License for more details.                         '
+#                                                                              '
+#  You should have received a copy of the GNU Lesser General Public License    '
+#  along with geoh5py.  If not, see <https://www.gnu.org/licenses/>.           '
+# ''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+from __future__ import annotations
+
+import numpy as np
+
+
+def normalize(v: np.ndarray) -> np.ndarray:
+    """
+    Normalize a vector to unit length.
+
+    :param v: Input vector to normalize.
+    :param dtype: NumPy data type for the normalized vector.
+
+    :return: Normalized vector with unit length.
+    """
+    v = np.asarray(v, dtype=np.float64)
+    return v / np.linalg.norm(v)
+
+
+def plane_from_quad_origin_last(
+    vertices: np.ndarray,
+) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Create a coordinate system from a quad with the last vertex as origin.
+
+    Constructs a plane coordinate system (origin, u-vector, v-vector, normal)
+    from four vertices where the last vertex serves as the origin.
+
+    :param vertices: Array of shape (4, 3) containing the four vertices of the quad.
+
+    :return: Tuple containing (origin, u-vector, v-vector, normal vector).
+
+    :raises ValueError: If vertices is not a (4, 3) array.
+    """
+    if vertices.shape != (4, 3):
+        raise ValueError("vertices must be a (4, 3) array")
+
+    vertices = np.asarray(vertices)
+
+    p0 = vertices[3]
+    u = normalize(vertices[2] - p0)
+    v = normalize(vertices[0] - p0)
+    n = normalize(np.cross(u, v))
+    return p0, u, v, n
+
+
+def intersect_segment_plane(
+    a: np.ndarray,
+    b: np.ndarray,
+    p0: np.ndarray,
+    n: np.ndarray,
```

Comment:
```suggestion
    start: np.ndarray,
    end: np.ndarray,
    origin: np.ndarray,
    normal: np.ndarray,
```
Same here, more descriptive variables are prefered